WCiOS: Update Fastfile with version model

spencertransier opened the pull request:
## Summary
- This PR integrates the new version and build code models from https://github.com/wordpress-mobile/release-toolkit/pull/512
- It currently points to a development version of the Release Toolkit so that I could test out the call site implementation before releasing a new version of the Release Toolkit. 
- Nearly all of the `precheck` actions from the Release Toolkit were tightly coupled with the existing version methods. To simplify things with the new version model, I took the relevant snippets from the precheck actions and added them directly here. This is something that I've discussed with Olivier and Jeremy before in the context of "no one knows what the precheck actions do unless you dive deep down into the RT". These changes are somewhat opposite of DRY, but I think it's worth it for transparency. We may also decide in the future to create smaller actions that do more specific things versus the existing precheck and version bump actions that combine many disparate methods. 

## Before Merging
- [x] Point the Release Toolkit to a stable version

spencertransier committed:
Begin using the Version model methods

spencertransier committed:
Finish replacing current methods with Version model

spencertransier committed:
Use development branch of the Release Toolkit

spencertransier committed:
Temporarily disable git clean and commit bump

spencertransier committed:
Use VERSION_FILE constant

spencertransier committed:
Fix typo and Rubocop violation

spencertransier committed:
Add additional methods for returning versions and build codes

spencertransier commented:
> Conflicting files
>
> Gemfile.lock

I will get this sorted out when there is a new Release Toolkit version to update to

spencertransier commented on fastlane/Fastfile:
```diff
@@ -1219,10 +1295,62 @@ def inject_buildkite_analytics_environment(xctestrun_path:)
 end
 
-# Returns the version of the app based on the specified xcconfig file
+
```
I created a test lane that verified these values are being calculated and formatted as expected (given a `VERSION_LONG` of `15.5.2` and a `VERSION_SHORT` of `15.5.2.4`):

<img width="377" alt="Screenshot 2023-09-28 at 11 35 30 AM" src="https://github.com/woocommerce/woocommerce-ios/assets/17955542/240e827c-d1eb-472d-b801-48ab4381efd8">

spencertransier commented on fastlane/Fastfile:
```diff
@@ -409,60 +534,6 @@ platform :ios do
-  # bundle exec fastlane build_and_upload_beta skip_confirm:true
-  #####################################################################################
-  desc 'Builds and uploads for distribution'
-  lane :build_and_upload_beta do |options|
```
After searching the release scenario, Buildkite scripts, and Buildkite pipelines, I did not find any instances of this lane being used.

spencertransier commented on fastlane/Fastfile:
```diff
@@ -409,60 +534,6 @@ platform :ios do
-  # bundle exec fastlane build_and_upload_release skip_confirm:true
-  #####################################################################################
-  desc 'Builds and uploads for distribution'
-  lane :build_and_upload_release do |options|
```
After searching the release scenario, Buildkite scripts, and Buildkite pipelines, I did not find any instances of this lane being used.

spencertransier commented on fastlane/Fastfile:
```diff
@@ -33,6 +34,12 @@ SCREENSHOT_DEVICES = [
   'iPad Pro (12.9-inch) (3rd generation)'
 ].freeze
 
+# Instanstiate versioning classes
+VERSION_CALCULATOR = Fastlane::Wpmreleasetoolkit::Versioning::MarketingVersionCalculator.new
+VERSION_FORMATTER = Fastlane::Wpmreleasetoolkit::Versioning::FourPartVersionFormatter.new
+BUILD_CODE_FORMATTER = Fastlane::Wpmreleasetoolkit::Versioning::FourPartBuildCodeFormatter.new
+VERSION_FILE = Fastlane::Wpmreleasetoolkit::Versioning::IOSVersionFile.new(xcconfig_path: PUBLIC_CONFIG_FILE)
```
This is where the various version and build code formatters/calculators are instantiated to be used by the rest of the Fastfile.

spencertransier committed:
Uncomment a Git checkout command and don't run the check for RT updates before every lane

spencertransier committed:
Fixed rubocop violations

spencertransier commented on fastlane/Fastfile:
```diff
@@ -135,9 +142,6 @@ before_all do |lane|
   # Skip these checks/steps for test lane (not needed for testing)
   next if lane == :test_without_building
 
-  # Ensure we use the latest version of the toolkit
-  check_for_toolkit_updates unless is_ci || ENV['FASTLANE_SKIP_TOOLKIT_UPDATE_CHECK']
-
```
I moved this check to the top of the `code_freeze` lane instead of `before_all`. It hasn't been useful to have it run before every single lane during the course of a release.

spencertransier committed:
Merge branch 'trunk' into update/release-toolkit-version-model

spencertransier committed:
Update Gemfile.lock

spencertransier committed:
Update Release Toolkit

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -277,17 +321,47 @@ platform :ios do
+    UI.user_error!("#{message}Release branch for version #{current_release_version} doesn't exist. Abort.") unless !options[:base_version].nil? || Fastlane::Helper::GitHelper.checkout_and_pull(release: current_release_version)
+
+    # Check user overwrite
+    switch_to_user_branch(options[:base_version], message) unless options[:base_version].nil?
```
~~I'm not sure I understand what this is supposed to do?~~

~~When  I look at the implementation for this method from what i can see it only does a check then adds a sentence to the `message`… but doesn't actually _git switch_ to a specific branch… So either the naming of that method is misleading, or it doesn't do what it's supposed to?~~

[EDIT] My bad, I now see that it in fact does `checkout_and_pull` after all, missed it in my initial read of that method's implementation. See https://github.com/woocommerce/woocommerce-ios/pull/10807#discussion_r1341722951 below.

spencertransier replied:
Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -277,17 +321,47 @@ platform :ios do
+
+    # Bump the build code
+    UI.message 'Bumping build code...'
+    Fastlane::Helper::GitHelper.ensure_on_branch!('release')
```
Fastlane has an action called `ensure_git_branch` that is built in, we should use this instead.

Also, one thing to know about is that our toolkit's `ensure_on_branch!` checks the branch name using `include?`; while the fastlane one (`ensure_git_branch`) does the check using `RegExp.match?`

This means that when you use `ensure_git_branch` you can even use `ensure_git_branch(branch: '^release/')` to check if the branch _starts_ with `release/`, which will match `release/12.3` but not `merge/release-into-trunk` for example.

spencertransier replied:
~Good to know! I'm pretty sure I spotted some uses of the built-in action. I'm going to hold off on changing all of these over to the built-in action for a separate PR. I've tried to just replicate everything that was in the precheck and version bump actions so that we reduce the possibility of introducing unexpected behavior with the rest of these changes.~

spencertransier replied:
Addressed here: https://github.com/woocommerce/woocommerce-ios/pull/10807/commits/b8802015356361954a34b2f3ea6e0241483966a7

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -307,11 +381,44 @@ platform :ios do
+    UI.message 'Bumping hotfix version and build code...'
+    VERSION_FILE.write(
+      version_short: new_version,
+      version_long: build_code(new_version)
```
I can't see a method named `build_code` defined anywhere? Was this supposed to be `BUILD_CODE_FORMATTER.build_code(version: new_version)` instead?

spencertransier replied:
That was left over from an earlier implementation of the build_code methods. I'll update it to `next_build_code` instead.

spencertransier replied:
Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

spencertransier committed:
Update Release Toolkit revision

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -542,7 +618,7 @@ platform :ios do
+    version = options[:beta_release] ? current_build_code : current_release_version
     create_release(
       repository: GITHUB_REPO,
       version:,
```
Missing a value for this parameter here 🤔

spencertransier replied:
That's the Ruby 3 convention of not needing to include the argument if the name of the argument is the same as the name of the parameter.

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -1240,6 +1367,11 @@ def prompt_user_for_app_store_connect_credentials
   ENV['FASTLANE_USER'] = CredentialsManager::AccountManager.new.user
 end
 
+def switch_to_user_branch(version, message)
+  UI.user_error!("Release branch for version #{version} doesn't exist. Abort.") unless Fastlane::Helper::GitHelper.checkout_and_pull(release: version)
+  message << "Looking at branch release/#{version} as requested by user. Detected version: #{current_build_code}.\n"
+end
```
Oh ok my bad, it does call `checkout_and_pull`, it's just that this is kinda hidden within the `unless` clause (I didn't expect any "action with side effect" to be in the `unless` clause, as opposed to just a boolean value, that's why I missed it)

Maybe to help readability we should rewrite this like that?
```suggestion
def switch_to_user_branch(version, message)
  success = Fastlane::Helper::GitHelper.checkout_and_pull(release: version)
  UI.user_error!("Release branch for version #{version} doesn't exist. Abort.") unless success
  message << "Looking at branch release/#{version} as requested by user. Detected version: #{current_build_code}.\n"
end
```

Also, I'm not a fan of having a method that modifies one of its input parameters (`message`), as opposed to return it as a new value. Doesn't feel very idiomatic of Ruby.

spencertransier replied:
Addressed in https://github.com/woocommerce/woocommerce-ios/commit/be0e00af7dcddbf962ccca507f0097815a128c46

spencertransier committed:
Update Release Toolkit to `9.1.0`

spencertransier committed:
Update `override_default_release_branch`

spencertransier committed:
Update minimum Release Toolkit to 9.1

spencertransier committed:
Move `ensure_git_status_clean` to top of lanes

spencertransier committed:
Fix typo and rubocop violations

spencertransier committed:
Add `hotfix_build_code` method

spencertransier committed:
Fix hotfix_build_code method

spencertransier committed:
Fix rubocop violation

spencertransier committed:
Use Fastlane's built-in `ensure_git_branch` action

spencertransier committed:
Use `ensure_git_branch` instead of `git_branch.start_with`

spencertransier committed:
Update for consistency across apps

mokagio reviewed (commented):
Just a note that I've been focusing on https://github.com/wordpress-mobile/WordPress-iOS/pull/21666 and leaving various suggestions there. 

I think it'd be best to lock in the implementation on that end, then propagate the learnings here.

spencertransier committed:
Revert Release Toolkit update change

spencertransier committed:
Improve precheck comments

spencertransier committed:
Improve user confirmation checks

spencertransier committed:
Update version and build code method names

spencertransier committed:
Remove three environment variables

spencertransier committed:
Update release notes constant name

spencertransier committed:
Remove unneeded `PUBLIC_CONFIG_FILE` env var

spencertransier committed:
Merge branch 'trunk' into update/release-toolkit-version-model

spencertransier committed:
Update Release Toolkit to `9.2.0`

spencertransier commented:
@mokagio I've updated this one with the changes from WP/JPiOS, so it is ready to review again.

spencertransier committed:
Merge pull request #10982 from woocommerce/remove/env-vars-from-fastfile

Remove unneeded env vars from Fastfile

mokagio commented on fastlane/Fastfile:
```diff
@@ -172,16 +176,49 @@ platform :ios do
+    MESSAGE
+
+    UI.important(message)
+    UI.user_error!('Aborted by user request') if !options[:skip_confirm] && !UI.confirm('Do you want to continue?')
```
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -172,16 +176,49 @@ platform :ios do
+
+    # Create the release branch
+    UI.message('Creating release branch...')
+    ensure_git_branch(branch: DEFAULT_BRANCH)
```
How do you feel about removing this check? What are the chances the branch changed under the hood after the `checkout_and_pull` command a few lines above?

```suggestion
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -172,16 +176,49 @@ platform :ios do
+      version_short: release_version_next,
+      version_long: build_code_code_freeze
+    )
+    Fastlane::Helper::Ios::GitHelper.commit_version_bump
```
I run into an issue with this during the 23.6 Jetpack/WordPress code freeze and ended up replacing it with `git_commit` in https://github.com/wordpress-mobile/WordPress-iOS/pull/21911/commits/f886436c4fcdbaf6d4fd3a1f65e06a3fc06b0b8e

More info in https://github.com/wordpress-mobile/release-toolkit/issues/525

How would you feel about doing the same here? Otherwise, because of the `ENV['PROJECT_ROOT_FOLDER']` above, I fear this will break on the next code freeze.

spencertransier replied:
Shoot, I missed that method during the env var project because it referenced the env var like `get_from_env!(key: 'PROJECT_ROOT_FOLDER')` instead of `ENV['PROJECT_ROOT_FOLDER']`. Sorry, I should have caught that. I'll make sure to get the other iOS apps updated. 

Fixed for this PR in https://github.com/woocommerce/woocommerce-ios/commit/97204c6a0430e90c290e9d72147eb9d993b8af4d

mokagio commented on fastlane/Fastfile:
```diff
@@ -277,17 +322,48 @@ platform :ios do
+    override_default_release_branch(options[:base_version]) unless options[:base_version].nil?
+
+    UI.important(message)
+    UI.user_error!('Aborted by user request') if !options[:skip_confirm] && !UI.confirm('Do you want to continue?')
```
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -307,11 +383,49 @@ platform :ios do
+    MESSAGE
+
+    UI.important(message)
+    UI.user_error!('Aborted by user request') if !options[:skip_confirm] && !UI.confirm('Do you want to continue?')
```
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -307,11 +383,49 @@ platform :ios do
+
+    # Create the hotfix branch
+    UI.message('Creating hotfix branch...')
+    # Verify that there's nothing in progress in the working copy
+    ensure_git_status_clean
```
In the interest of failing early, should we move this at the start of the lane?

spencertransier replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/commit/01da9abb2de7063634358f8f19c9f89a0009376a

mokagio commented on fastlane/Fastfile:
```diff
@@ -320,24 +434,19 @@ platform :ios do
-      bypass: ENV.fetch('RELEASE_TOOLKIT_SKIP_PUSH_CONFIRM', nil)
-    )
-      push_to_git_remote(tags: false)
-      trigger_release_build(branch_to_build: "release/#{app_version}")
-    else
-      UI.message('Aborting hotfix finalization. See you later 👋')
-    end
+    # Verify that there's nothing in progress in the working copy
+    ensure_git_status_clean
+
+    UI.important("Triggering hotfix build for version: #{release_version_current}")
+    UI.user_error!('Aborted by user request') if !options[:skip_confirm] && !UI.confirm('Do you want to continue?')
+    trigger_release_build(branch_to_build: "release/#{release_version_current}")
```
I don't remember where we discussed this already, but I if I recall correctly the conclusion was that the push command should have remained in the lane?

Otherwise, we might run into a scenario where the release manager has changes locally that they assume would be pushed (like the other finalization lanes do) but the CI build is triggered on the latest commit on the remote.

Or... did we conclude that there should be no local change because all changes in hotfixes happen via PRs and are already in the remote?

spencertransier replied:
That's a great point. I don't think I was a part of that discussion, but I think that scenario makes sense. _Most likely_, all the changes will be done as PRs, but might as well push any local changes just in case. 

I added the push command in https://github.com/woocommerce/woocommerce-ios/commit/8a70f61035d4d9de7c120791bbfb210e1d716f62

spencertransier committed:
Use positive logic

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

spencertransier committed:
Remove unneeded branch check

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

spencertransier committed:
Move git clean check to beginning of lane

spencertransier committed:
Use positive logic

spencertransier committed:
Push local changes during hotfix finalization

spencertransier committed:
Remove extra parentheses

spencertransier committed:
Add `commit_version_bump` method to Fastfile

spencertransier commented on fastlane/Fastfile:
```diff
@@ -510,17 +577,19 @@ platform :ios do
+    ensure_sentry_installed
+
+    UI.important("Building version #{release_version_current} (#{build_code_current}) and uploading to TestFlight")
+    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?'))
```
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/10807/commits/5cd92da003c3a407bcd0c675b63c72db14c8c87e

spencertransier commented:
@mokagio Thank you for taking another look! Sorry, I thought I had already fixed most of those cases for WCiOS, but I think I was getting it mixed up with another app 🤪. All your comments are now addressed 👍

mokagio reviewed (approved):
All looks good. At this point, the best thing to do is to merge and see how we go on the next code freeze 👍
